File: app/src/main/java/vng/zalo/tdtai/zalo/zalo/views/home/fragments/chat_fragment/ChatFragmentAdapter.java
package vng.zalo.tdtai.zalo.zalo.views.home.fragments.chat_fragment;

import android.graphics.Typeface;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.ListAdapter;
import androidx.recyclerview.widget.RecyclerView;

import com.squareup.picasso.Picasso;

import de.hdodenhof.circleimageview.CircleImageView;
import vng.zalo.tdtai.zalo.R;
import vng.zalo.tdtai.zalo.zalo.models.RoomItem;
import vng.zalo.tdtai.zalo.zalo.utils.Constants;
import vng.zalo.tdtai.zalo.zalo.utils.ModelViewHolder;
import vng.zalo.tdtai.zalo.zalo.utils.Utils;

public class ChatFragmentAdapter extends ListAdapter<RoomItem, ChatFragmentAdapter.ChatViewHolder> {
    private static final String TAG = ChatFragmentAdapter.class.getSimpleName();

    private ChatFragment chatFragment;

    ChatFragmentAdapter(ChatFragment chatFragment, @NonNull DiffUtil.ItemCallback<RoomItem> diffCallback) {
        super(diffCallback);
        this.chatFragment = chatFragment;
    }

    @NonNull
    @Override
    public ChatViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_chat_fragment, parent, false);
        return new ChatViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull ChatViewHolder holder, int position) {
        holder.bind(position);
    }

    @Override
    public int getItemCount() {
        return getCurrentList().size();
    }

    class ChatViewHolder extends RecyclerView.ViewHolder implements ModelViewHolder {
        View itemView;
        TextView nameTextView;
        TextView descTextView;
        TextView timeTextView;
        TextView iconTextView;
        CircleImageView avatarImgView;

        ChatViewHolder(@NonNull View itemView) {
            super(itemView);
            this.itemView = itemView;
            nameTextView = itemView.findViewById(R.id.nameTextView);
            descTextView = itemView.findViewById(R.id.descTextView);
            timeTextView = itemView.findViewById(R.id.receiveTimeTextView);
            iconTextView = itemView.findViewById(R.id.iconTextView);
            avatarImgView = itemView.findViewById(R.id.avatarImgView);
        }

        @Override
        public void bind(int position) {
            itemView.setOnClickListener(chatFragment);
            RoomItem roomItem = getItem(position);

            nameTextView.setText(roomItem.name);
            Utils.formatTextOnNumberOfLines(nameTextView, 1);

            if (roomItem.lastMsgTime != null) {
                String formatTime = Utils.getTimeDiffOrFormatTime(roomItem.lastMsgTime.toDate());
                timeTextView.setText(formatTime);
            } else {
                timeTextView.setText("");
            }

            descTextView.setText(roomItem.lastMsg);
            Utils.formatTextOnNumberOfLines(descTextView, 1);

            Picasso.with(avatarImgView.getContext())
                    .load(roomItem.avatar)
                    .fit()
                    .into(avatarImgView);

            if (roomItem.unseenMsgNum > 0) {
                String unseenMsgNum = roomItem.unseenMsgNum < Constants.MAX_UNSEEN_MSG_NUM ?
                        roomItem.unseenMsgNum.toString() : (Constants.MAX_UNSEEN_MSG_NUM + "+");
                iconTextView.setText(unseenMsgNum);
                iconTextView.setVisibility(View.VISIBLE);
                nameTextView.setTypeface(null, Typeface.BOLD);
                descTextView.setTypeface(null, Typeface.BOLD);
                timeTextView.setTypeface(null, Typeface.BOLD);
            } else {
                iconTextView.setVisibility(View.GONE);
                nameTextView.setTypeface(null, Typeface.NORMAL);
                descTextView.setTypeface(null, Typeface.NORMAL);
                timeTextView.setTypeface(null, Typeface.NORMAL);
            }
        }
    }
}

File: app/src/main/java/vng/zalo/tdtai/zalo/ui/home/chat/ChatFragmentAdapter.kt
package vng.zalo.tdtai.zalo.ui.home.chat

import android.graphics.Typeface
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.core.content.ContextCompat
import com.squareup.picasso.Picasso
import kotlinx.android.synthetic.main.item_room.view.*
import vng.zalo.tdtai.zalo.R
import vng.zalo.tdtai.zalo.base.BaseListAdapter
import vng.zalo.tdtai.zalo.base.BindableViewHolder
import vng.zalo.tdtai.zalo.data_model.room.Room
import vng.zalo.tdtai.zalo.data_model.room.RoomItem
import vng.zalo.tdtai.zalo.data_model.room.RoomItemPeer
import vng.zalo.tdtai.zalo.manager.ResourceManager
import vng.zalo.tdtai.zalo.manager.SessionManager
import vng.zalo.tdtai.zalo.util.Constants
import vng.zalo.tdtai.zalo.util.RoomItemDiffCallback
import vng.zalo.tdtai.zalo.util.Utils
import vng.zalo.tdtai.zalo.util.smartLoad
import javax.inject.Inject

class ChatFragmentAdapter @Inject constructor(
        private val clickListener: View.OnClickListener,
        private val sessionManager: SessionManager,
        private val resourceManager: ResourceManager,
        private val utils: Utils,
        diffCallback: RoomItemDiffCallback) : BaseListAdapter<RoomItem, ChatFragmentAdapter.RoomItemViewHolder>(diffCallback) {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RoomItemViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.item_room, parent, false)
        view.setOnClickListener(clickListener)
        return RoomItemViewHolder(view)
    }

    override fun onBindViewHolder(holder: RoomItemViewHolder, position: Int) {
        holder.bind(position)
    }

    override fun onBindViewHolder(
            holder: RoomItemViewHolder,
            position: Int,
            payloads: ArrayList<*>
    ) {
        val roomItem = currentList[position]
        payloads.forEach {
            when (it) {
                RoomItem.PAYLOAD_NEW_MESSAGE -> holder.bindLastMessage(roomItem)
                RoomItem.PAYLOAD_SEEN_MESSAGE -> holder.bindSeenStatus(roomItem)
                RoomItem.PAYLOAD_ONLINE_STATUS -> holder.bindOnlineStatus(roomItem)
                RoomItem.PAYLOAD_TYPING -> holder.bindLastMessage(roomItem)
            }
        }
    }

    inner class RoomItemViewHolder(itemView: View) : BindableViewHolder(itemView) {
        override fun bind(position: Int) {
            itemView.apply {
                val roomItem = getItem(position)

                nameTextView.text = roomItem.getDisplayName(resourceManager)

                Picasso.get()
                        .smartLoad(roomItem.avatarUrl, resourceManager, watchOwnerAvatarImgView) {
                            it.placeholder(if (roomItem.roomType == Room.TYPE_PEER) R.drawable.default_peer_avatar else R.drawable.default_group_avatar)
                                    .fit()
                                    .centerCrop()
                        }

                bindLastMessage(roomItem)
                bindSeenStatus(roomItem)
                bindOnlineStatus(roomItem)
            }
        }

        fun bindOnlineStatus(roomItem: RoomItem) {
            if (roomItem.roomType == Room.TYPE_PEER) {
                itemView.apply {
                    if ((roomItem as RoomItemPeer).lastOnlineTime == null) {
                        onlineStatusImgView.visibility = View.VISIBLE
                    } else {
                        onlineStatusImgView.visibility = View.GONE
                    }
                }
            }
        }

        fun bindLastMessage(roomItem: RoomItem) {
            itemView.apply {
                timeTextView.text = if (roomItem.lastMsgTime != null) utils.getTimeDiffFormat(roomItem.lastMsgTime!!) else ""

                descTextView.text = if (roomItem.lastTypingMember != null) {
                    typingAnimView.visibility = View.VISIBLE
                    typingAnimView.postDelayed({
                        typingAnimView.playAnimation()
                    }, 100)
                    descTextView.setTextColor(ContextCompat.getColor(context, R.color.lightPrimary))
                    String.format("%s: %s",
                            if (roomItem.lastTypingMember!!.userId == sessionManager.curUser!!.id) {
                                context.getString(R.string.label_me)
                            } else {
                                Picasso.get()
                                        .smartLoad(roomItem.lastTypingMember!!.avatarUrl, resourceManager, typingUserAvatarImgView) {
                                            it.placeholder(R.drawable.default_peer_avatar)
                                                    .fit()
                                                    .centerCrop()
                                        }

                                typingUserAvatarImgView.visibility = View.VISIBLE

                                resourceManager.getNameFromId(roomItem.lastTypingMember!!.userId!!)
                                        ?: roomItem.lastTypingMember!!.name
                            },
                            context.getString(R.string.description_typing)
                    )
                } else {
                    typingUserAvatarImgView.visibility = View.GONE

                    typingAnimView.visibility = View.GONE
                    descTextView.setTextColor(ContextCompat.getColor(context, android.R.color.black))
                    if (roomItem.lastSenderId != null) {
                        val senderName =
                                if (roomItem.lastSenderId == sessionManager.curUser!!.id)
                                    context.getString(R.string.label_me)
                                else
                                    resourceManager.getNameFromId(roomItem.lastSenderId!!)
                                            ?: roomItem.lastSenderName

                        String.format("%s: %s", senderName, roomItem.lastMsg)
                    } else ""
                }
            }
        }

        fun bindSeenStatus(roomItem: RoomItem) {
            itemView.apply {
                if (roomItem.unseenMsgNum > 0) {
                    // if unseenMsgNum > N -> display: [N+]
                    iconTextView.text =
                            if (roomItem.unseenMsgNum < Constants.MAX_UNSEEN_MSG_NUM)
                                roomItem.unseenMsgNum.toString()
                            else
                                "${Constants.MAX_UNSEEN_MSG_NUM}+"

                    adjustViewsBasedOnSeenStatus(false)
                } else {
                    adjustViewsBasedOnSeenStatus(true)
                }
            }
        }

        private fun adjustViewsBasedOnSeenStatus(isSeen: Boolean) {
            itemView.apply {
                if (isSeen) {
                    iconTextView.visibility = View.GONE
                    descTextView.setTypeface(null, Typeface.NORMAL)
                    timeTextView.setTypeface(null, Typeface.NORMAL)
                } else {
                    iconTextView.visibility = View.VISIBLE
                    descTextView.setTypeface(null, Typeface.BOLD)
                    timeTextView.setTypeface(null, Typeface.BOLD)
                }
            }
        }
    }
}